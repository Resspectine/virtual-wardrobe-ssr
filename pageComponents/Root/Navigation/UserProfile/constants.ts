export const placeholderImageUrl = 'https://via.placeholder.com/150';
