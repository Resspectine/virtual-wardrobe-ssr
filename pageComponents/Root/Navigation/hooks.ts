import { useRouter } from 'next/router';

export const useNavigationControl = () => {
  const { route: appRoute, push } = useRouter();

  const goTo = (link: string): void => {
    push(link);
  };

  const isCurrentPage = (route: string): boolean => appRoute.includes(route);

  return {
    isCurrentPage,
    goTo,
  };
};
