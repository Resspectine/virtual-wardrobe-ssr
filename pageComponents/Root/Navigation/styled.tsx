import { Typography } from '@mui/material';
import { Box, styled } from '@mui/system';

export const NavigationWrapper = styled(Box)(({ theme }) => ({
  padding: '20px',
  backgroundColor: theme.palette.background.paper,
  display: 'flex',
  gap: 20,
  justifyContent: 'center',
  position: 'relative',
  alignItems: 'center',
}));

export const NavigationLink = styled(Typography, {
  shouldForwardProp: propName => propName !== 'isCurrentPage',
})<{ isCurrentPage: boolean }>(({ theme, isCurrentPage }) => ({
  textDecoration: 'none',
  color: isCurrentPage ? theme.palette.text.secondary : theme.palette.text.primary,
  fontSize: 18,
  cursor: 'pointer',
}));
