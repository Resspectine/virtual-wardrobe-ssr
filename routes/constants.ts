export const ROUTE_PATHS = {
  login: '/login',
  register: '/register',
  main: '/main',
  profile: '/profile',
  createGarment: '/create/garment',
  editGarment: (id = '[id]'): string => `/create/garment/${id}`,
  404: '/*',
};

export const KEYS_TO_OMIT = ['editGarment', 404, 'login', 'register', 'profile'] as const;

export type TypesToOmit = typeof KEYS_TO_OMIT[number];
