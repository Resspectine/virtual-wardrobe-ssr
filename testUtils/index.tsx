import { queries, Queries, render as reactRender, RenderOptions, RenderResult } from '@testing-library/react';
import { FC, ReactElement } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

import { ThemeProvider } from '@/materialUiTheme/provider';

export const sleep = (timeout: number): Promise<void> =>
  new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, timeout);
  });

export const render = <Q extends Queries = typeof queries, Container extends Element | DocumentFragment = HTMLElement>(
  ui: ReactElement,
  options: RenderOptions<Q, Container> = {}
): RenderResult<Q, Container> => {
  const queryClient = new QueryClient();

  const Wrapper: FC = ({ children }) => (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>{children}</ThemeProvider>
    </QueryClientProvider>
  );

  return { ...reactRender(ui, { ...options, wrapper: Wrapper }) };
};
