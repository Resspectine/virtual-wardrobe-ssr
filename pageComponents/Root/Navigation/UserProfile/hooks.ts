import { useRouter } from 'next/router';
import { useState, MouseEventHandler } from 'react';
import { useMutation } from 'react-query';

import { placeholderImageUrl } from './constants';

import { appFetch } from '@/lib/controller';
import { usePictureLoad } from '@/lib/hooks/usePictureLoad';
import { ROUTE_PATHS } from '@/routes/constants';
import { useAppState } from '@/store/appState';
import { useUser } from '@/store/user';

export const useUserProfile = () => {
  const router = useRouter();
  const user = useUser(state => state.user);
  const removeUser = useUser(state => state.removeUser);
  const userProfileUrl = usePictureLoad({ file: user?.avatar });
  const toggleThemeMode = useAppState(state => state.toggleThemeMode);
  const [anchorEl, setAnchorEl] = useState<(EventTarget & HTMLButtonElement) | null>(null);

  const { mutate: logOutMutate } = useMutation(() => appFetch('authentication/log-out', { method: 'POST' }), {
    onMutate: () => {
      removeUser();
    },
  });

  const handleClick: MouseEventHandler<HTMLButtonElement> = event => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = (): void => {
    setAnchorEl(null);
  };

  const open = !!anchorEl;

  const menuItems = [
    {
      text: 'Profile',
      onClick: (): void => {
        handleClose();

        router.push(ROUTE_PATHS.profile);
      },
    },
    {
      text: 'Toggle theme',
      onClick: (): void => {
        handleClose();
        toggleThemeMode();
      },
    },
    {
      text: 'Logout',
      onClick: (): void => {
        logOutMutate();
        handleClose();
      },
    },
  ];

  return {
    open,
    anchorEl,
    menuItems,
    handleClick,
    handleClose,
    userProfileUrl: userProfileUrl || placeholderImageUrl,
  };
};
