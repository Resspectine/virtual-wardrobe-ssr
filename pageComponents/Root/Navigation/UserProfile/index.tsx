import { Avatar, Menu, MenuItem } from '@mui/material';
import { FC } from 'react';

import { useUserProfile } from './hooks';
import { UserProfileButton, UserProfileWrapper } from './styled';

const UserProfile: FC = () => {
  const { open, anchorEl, menuItems, handleClick, handleClose, userProfileUrl } = useUserProfile();

  return (
    <UserProfileWrapper>
      <UserProfileButton onClick={handleClick}>
        <Avatar src={userProfileUrl} alt="avatar" />
      </UserProfileButton>
      <Menu anchorEl={anchorEl} open={open} onClose={handleClose}>
        {menuItems.map(({ text, onClick }, index) => (
          <MenuItem key={index} onClick={onClick}>
            {text}
          </MenuItem>
        ))}
      </Menu>
    </UserProfileWrapper>
  );
};

export default UserProfile;
