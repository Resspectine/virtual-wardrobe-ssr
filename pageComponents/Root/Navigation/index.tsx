import Box from '@mui/material/Box';
import omit from 'lodash/omit';
import { FC } from 'react';

import { useNavigationControl } from './hooks';
import { NavigationLink, NavigationWrapper } from './styled';
import UserProfile from './UserProfile';

import { camelToSentenceCase } from '@/lib/helpers/textTransformations';
import { KEYS_TO_OMIT, ROUTE_PATHS, TypesToOmit } from '@/routes/constants';

const routes = omit(ROUTE_PATHS, KEYS_TO_OMIT) as Omit<typeof ROUTE_PATHS, TypesToOmit>;

const Navigation: FC = () => {
  const { goTo, isCurrentPage } = useNavigationControl();

  return (
    <NavigationWrapper>
      {Object.entries(routes).map(([linkName, link], index) => (
        <Box key={index} onClick={(): void => goTo(link)}>
          <NavigationLink isCurrentPage={isCurrentPage(link)}>{camelToSentenceCase(linkName)}</NavigationLink>
        </Box>
      ))}
      <UserProfile />
    </NavigationWrapper>
  );
};

export default Navigation;
