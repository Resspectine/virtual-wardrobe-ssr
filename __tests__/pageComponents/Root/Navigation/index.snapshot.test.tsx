import { NextRouter, useRouter } from 'next/router';

import Navigation from '@/pageComponents/Root/Navigation';
import UserProfile from '@/pageComponents/Root/Navigation/UserProfile';
import { ROUTE_PATHS } from '@/routes/constants';
import { render } from '@/testUtils';

jest.mock('next/router');
jest.mock('@/pageComponents/Root/Navigation/UserProfile');

const mockedUseRouter = useRouter as unknown as jest.Mock<ReturnType<typeof useRouter>>;
const mockedUserProfile = UserProfile as unknown as jest.Mock<ReturnType<typeof UserProfile>>;

afterEach(() => {
  mockedUseRouter.mockRestore();
  mockedUserProfile.mockRestore();
});

describe('Navigation', () => {
  test('should match snapshot on main page', async () => {
    const push = jest.fn(() => {});
    mockedUseRouter.mockImplementation(
      () =>
        ({
          route: ROUTE_PATHS.main,
          push,
        } as unknown as NextRouter)
    );
    mockedUserProfile.mockImplementation(() => null);

    const { asFragment } = render(<Navigation />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should match snapshot on create garment page', async () => {
    const push = jest.fn(() => {});
    mockedUseRouter.mockImplementation(
      () =>
        ({
          route: ROUTE_PATHS.createGarment,
          push,
        } as unknown as NextRouter)
    );
    mockedUserProfile.mockImplementation(() => null);

    const { asFragment } = render(<Navigation />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should match snapshot on edit garment page', async () => {
    const push = jest.fn(() => {});
    mockedUseRouter.mockImplementation(
      () =>
        ({
          route: ROUTE_PATHS.editGarment(),
          push,
        } as unknown as NextRouter)
    );
    mockedUserProfile.mockImplementation(() => null);

    const { asFragment } = render(<Navigation />);

    expect(asFragment()).toMatchSnapshot();
  });
});
