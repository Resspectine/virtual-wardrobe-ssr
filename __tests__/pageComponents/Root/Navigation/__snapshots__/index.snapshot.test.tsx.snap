// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should match snapshot on create garment page 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 20px;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 20px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  margin: 0;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgba(0, 0, 0, 0.87);
  font-size: 18px;
  cursor: pointer;
}

.emotion-4 {
  margin: 0;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #388e3c;
  font-size: 18px;
  cursor: pointer;
}

<div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-2"
      >
        Main
      </p>
    </div>
    <div
      class="MuiBox-root emotion-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-4"
      >
        Create Garment
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navigation should match snapshot on edit garment page 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 20px;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 20px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  margin: 0;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgba(0, 0, 0, 0.87);
  font-size: 18px;
  cursor: pointer;
}

.emotion-4 {
  margin: 0;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #388e3c;
  font-size: 18px;
  cursor: pointer;
}

<div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-2"
      >
        Main
      </p>
    </div>
    <div
      class="MuiBox-root emotion-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-4"
      >
        Create Garment
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navigation should match snapshot on main page 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 20px;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 20px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  margin: 0;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #388e3c;
  font-size: 18px;
  cursor: pointer;
}

.emotion-4 {
  margin: 0;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgba(0, 0, 0, 0.87);
  font-size: 18px;
  cursor: pointer;
}

<div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-2"
      >
        Main
      </p>
    </div>
    <div
      class="MuiBox-root emotion-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-4"
      >
        Create Garment
      </p>
    </div>
  </div>
</DocumentFragment>
`;
